### PR TITLE
#489 read children in TkClaim + puzzle

### PR DESCRIPTION
**Pull request by amihaiemil**
Pr for #489 

Inside TkClaim its children are also read from the Footprint (the Claims which have the ``cause`` element equal to this claim's id).

Left a puzzle for editing ``claim.xsl`` in such a way that its children are also displayed.

### CODE REVIEW

File: src/main/java/com/zerocracy/tk/project/TkClaim.java [hunk 1]
```diff
@@ -69,6 +81,27 @@ public Response act(final RqRegex request) throws IOException {
         final String user = new RqUser(this.farm, request).value();
         final long cid = Long.valueOf(request.matcher().group(2));
         try (final Footprint ftp = new Footprint(this.farm, pkt)) {
+            final List<XeSource> children = new ArrayList<>(10);
+            for (
```
Comment (amihaiemil) [inline src/main/java/com/zerocracy/tk/project/TkClaim.java]: Please let me know if there is a shorter way of doing this iteration and addition to ``children`` List, I am not so experienced in Java 8.

**Comment by 0crat** [comment]
Job [#704](https://github.com/zerocracy/farm/issues/704) is now in scope, role is `REV`

File: src/main/java/com/zerocracy/tk/project/TkClaim.java [hunk 2]
```diff
@@ -69,6 +81,27 @@ public Response act(final RqRegex request) throws IOException {
...
+                final Document child : ftp.collection().find(
+                    Filters.eq("cause", cid)
```
Comment (amihaiemil) [inline src/main/java/com/zerocracy/tk/project/TkClaim.java]: Here, I assumed that ``cause`` is a simple element that should be equal to a claim's id.

**Comment by 0crat** [comment]
This pull request [#704](https://github.com/zerocracy/farm/issues/704) is assigned to @carlosmiranda[/z](https://www.0crat.com/u/carlosmiranda), here is [why](https://www.0crat.com/footprint/C3NDPUA8L/611805901). The budget is 15 minutes, see [§4](http://www.zerocracy.com/policy.html#4). Please, read [§27](http://www.zerocracy.com/policy.html#27) and when you decide to accept the changes, inform @yegor256[/z](https://www.0crat.com/u/yegor256) (the architect) right in this ticket. If you decide that this PR should not be accepted ever, also inform the architect.

**Comment by amihaiemil** [comment]
@carlosmiranda don't forget to review this one pls :)

**Comment by amihaiemil** [comment]
@carlosmiranda ping

**Comment by amihaiemil** [comment]
@carlosmiranda done, thanks :)

**Comment by carlosmiranda** [comment]
@yegor256 good to merge

**Comment by amihaiemil** [comment]
@yegor256 ping :)

**Comment by yegor256** [comment]
@amihaiemil why no unit testing? We have `TkClaimTest`.

**Comment by amihaiemil** [comment]
@yegor256 Right, I'll have a look into it :)

**Comment by amihaiemil** [comment]
@yegor256 pls read what I wrote in the description of this PR: is my assumption correct, that a claim should have a ``cause`` element (node) pointing to its parent id?

**Comment by amihaiemil** [comment]
@yegor256 I opened https://github.com/zerocracy/farm/issues/756 for documenting Claims in more detail. Since it doesn't make sense to hold the Issue and this Pr for more tests (the UI is also puzzled out), I left a puzzle for writing more test cases -- I added a check in the current test, for empty children element.

**Comment by yegor256** [comment]
@rultor merge

**Comment by yegor256** [comment]
@amihaiemil in the future, remember that you should start with a test and puzzle the code, when you have no time. Not the other way around. See this and this: http://www.yegor256.com/2017/03/24/tdd-that-works.html http://www.yegor256.com/2015/02/12/top-down-design.html

**Comment by amihaiemil** [comment]
@yegor256 Thanks, I'll remember that. Can you try to merge now, I've updated the puzzle. After it's merged, pls close the original Issue (#489), you are its author :D

**Comment by yegor256** [comment]
@rultor merge

**Comment by 0crat** [comment]
@ypshenychka[/z](https://www.0crat.com/u/ypshenychka) please review this job completed by @carlosmiranda[/z](https://www.0crat.com/u/carlosmiranda), as in [§30](http://www.zerocracy.com/policy.html#30); the job will be fully closed and all payments will be made when the quality review is completed

**Comment by 0crat** [comment]
Payment to `ARC` for a closed pull request, as in [§28](http://www.zerocracy.com/policy.html#28): +10 points just awarded to @yegor256[/z](https://www.0crat.com/u/yegor256)

**Comment by 0crat** [comment]
This pull request [#704](https://github.com/zerocracy/farm/issues/704) is assigned to @carlosmiranda[/z](https://www.0crat.com/u/carlosmiranda), here is [why](https://www.0crat.com/footprint/C3NDPUA8L/670813081). The budget is 15 minutes, see [§4](http://www.zerocracy.com/policy.html#4). Please, read [§27](http://www.zerocracy.com/policy.html#27) and when you decide to accept the changes, inform @yegor256[/z](https://www.0crat.com/u/yegor256) (the architect) right in this ticket. If you decide that this PR should not be accepted ever, also inform the architect.

**Comment by ypshenychka** [comment]
@amihaiemil According to our [QA Rules](http://www.zerocracy.com/policy.html#42):
>Messages in a job [always](http://www.yegor256.com/2014/11/24/principles-of-bug-tracking.html#4-avoid-noiseaddress-your-comments) start with a name of a user they are addressed to.

Please correct your code review comments by indicating an addressee in the beginning.

**Comment by ypshenychka** [comment]
@carlosmiranda According to our [QA Rules](http://www.zerocracy.com/policy.html#42):
>The code reviewer found at least three problems in the code.
Comments were mostly about design problems, not cosmetic issues.

Only one issue was found during code review. 
Please confirm that you'll try to find at least three problems while future reviews.

**Comment by amihaiemil** [comment]
@ypshenychka done, I updated by review comments :)

**Comment by ypshenychka** [comment]
@amihaiemil great, thanks!

**Comment by carlosmiranda** [comment]
@ypshenychka this is confirmed

**Comment by ypshenychka** [comment]
@carlosmiranda thanks

**Comment by ypshenychka** [comment]
@0crat quality acceptable

**Comment by 0crat** [comment]
Order was finished, quality is "acceptable": +15 points just awarded to @carlosmiranda[/z](https://www.0crat.com/u/carlosmiranda)

**Comment by 0crat** [comment]
`QA` review completed: +8 points just awarded to @ypshenychka[/z](https://www.0crat.com/u/ypshenychka)

**Comment by 0crat** [comment]
@ypshenychka[/z](https://www.0crat.com/u/ypshenychka) please review this job completed by @carlosmiranda[/z](https://www.0crat.com/u/carlosmiranda), as in [§30](http://www.zerocracy.com/policy.html#30); the job will be fully closed and all payments will be made when the quality review is completed

**Comment by 0crat** [comment]
The job [#704](https://github.com/zerocracy/farm/issues/704) is now out of scope

**Comment by 0crat** [comment]
Payment to `ARC` for a closed pull request, as in [§28](http://www.zerocracy.com/policy.html#28): +10 points just awarded to @yegor256[/z](https://www.0crat.com/u/yegor256)

**Comment by ypshenychka** [comment]
@yegor256 Here also QA review was already done.

**Comment by yegor256** [comment]
@0crat status

**Comment by 0crat** [comment]
> @0crat status ([here](https://github.com/zerocracy/farm/issues/704#issuecomment-374215551))

@yegor256 This is what I know about this job, as in [§32](http://www.zerocracy.com/policy.html#32):

  * The job [#704](https://github.com/zerocracy/farm/issues/704) is not in scope
  * The job is waiting quality review verdict by @ypshenychka[/z](https://www.0crat.com/u/ypshenychka)
  * Job [footprint](https://www.0crat.com/footprint/C3NDPUA8L?q=%7Bjob%3A%27gh%3Azerocracy%2Ffarm%23704%27%7D) (restricted area)

**Comment by yegor256** [comment]
@0crat status

**Comment by 0crat** [comment]
> @0crat status ([here](https://github.com/zerocracy/farm/issues/704#issuecomment-374242518))

@yegor256 There is an unrecoverable failure on my side. Please, submit it [here](https://github.com/zerocracy/farm/issues):

```
com.zerocracy.farm.reactive.StkGroovy[125] java.lang.IllegalStateException: java.lang.reflect.InvocationTargetException in com/zerocracy/stk/pm/check_job_status.groovy
sun.reflect.GeneratedMethodAccessor30[-1] java.lang.reflect.InvocationTargetException: null
com.jcabi.log.PreFormatter[141] java.lang.ArrayIndexOutOfBoundsException: 1
```

[1.0-SNAPSHOT](https://github.com/zerocracy/farm)

**Comment by yegor256** [comment]
@0crat status

**Comment by 0crat** [comment]
> @0crat status ([here](https://github.com/zerocracy/farm/issues/704#issuecomment-374254599))

@yegor256 There is an unrecoverable failure on my side. Please, submit it [here](https://github.com/zerocracy/farm/issues):

```
com.zerocracy.farm.reactive.StkGroovy[125] java.lang.IllegalStateException: java.lang.reflect.InvocationTargetException in com/zerocracy/stk/pm/check_job_status.groovy
sun.reflect.NativeMethodAccessorImpl[-2] java.lang.reflect.InvocationTargetException: null
com.jcabi.log.PreFormatter[141] java.lang.ArrayIndexOutOfBoundsException: 1
```

[1.0-SNAPSHOT](https://github.com/zerocracy/farm)

**Comment by yegor256** [comment]
@0crat status

**Comment by 0crat** [comment]
> @0crat status ([here](https://github.com/zerocracy/farm/issues/704#issuecomment-374261348))

@yegor256 There is an unrecoverable failure on my side. Please, submit it [here](https://github.com/zerocracy/farm/issues):

```
com.zerocracy.farm.reactive.StkGroovy[125] java.lang.IllegalStateException: java.lang.reflect.InvocationTargetException in com/zerocracy/stk/pm/check_job_status.groovy
sun.reflect.NativeMethodAccessorImpl[-2] java.lang.reflect.InvocationTargetException: null
com.jcabi.log.PreFormatter[141] java.lang.ArrayIndexOutOfBoundsException: 1
```

[1.0-SNAPSHOT](https://github.com/zerocracy/farm)

**Comment by yegor256** [comment]
@0crat status

**Comment by 0crat** [comment]
> @0crat status ([here](https://github.com/zerocracy/farm/issues/704#issuecomment-374263891))

@yegor256 This is what I know about this job, as in [§32](http://www.zerocracy.com/policy.html#32):

  * The job [#704](https://github.com/zerocracy/farm/issues/704) is not in scope
  * The job is waiting quality review verdict by @ypshenychka[/z](https://www.0crat.com/u/ypshenychka) for 2days
  * @carlosmiranda[/z](https://www.0crat.com/u/carlosmiranda) will be paid 15 minutes after quality review
  * Job [footprint](https://www.0crat.com/footprint/C3NDPUA8L?q=%7Bjob%3A%27gh%3Azerocracy%2Ffarm%23704%27%2Ctype%3A%7B%24not%3A%2F%5ENotify%2F%7D%7D) (restricted area)

**Comment by yegor256** [comment]
@0crat quality bad

**Comment by yegor256** [comment]
@ypshenychka it's a bug, ignore it

**Comment by 0crat** [comment]
Quality is low, no payment, see [§31](http://www.zerocracy.com/policy.html#31)

**Comment by 0crat** [comment]
Quality review completed: +8 points just awarded to @yegor256[/z](https://www.0crat.com/u/yegor256)
